pkg/comp-functions/functions/vshn-postgres-func/extensions.go

```diff
@@ -0,0 +1,96 @@
+package vshnpostgres
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	stackgresv1 "github.com/vshn/appcat/apis/stackgres/v1"
+	vshnv1 "github.com/vshn/appcat/apis/vshn/v1"
+	"github.com/vshn/appcat/pkg/comp-functions/runtime"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	timescaleExtName   = "timescaledb"
+	configResourceName = "pg-conf"
+	sharedLibraries    = "shared_preload_libraries"
+)
+
+// AddExtensions adds the user specified extensions to the SGCluster.
+func AddExtensions(ctx context.Context, iof *runtime.Runtime) runtime.Result {
+
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Starting extensions function")
+
+	comp := &vshnv1.VSHNPostgreSQL{}
+	err := iof.Observed.GetComposite(ctx, comp)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "Cannot get composite from function io", err)
+	}
+
+	extensionsMap := map[string]stackgresv1.SGClusterSpecPostgresExtensionsItem{}
+
+	for _, ext := range comp.Spec.Parameters.Service.Extensions {
+		extensionsMap[ext.Name] = stackgresv1.SGClusterSpecPostgresExtensionsItem{
+			Name: ext.Name,
+		}
+	}
+
+	// ensure pg_repack is always enabled
+	extensionsMap["pg_repack"] = stackgresv1.SGClusterSpecPostgresExtensionsItem{
+		Name: "pg_repack",
+	}
+
+	if _, ok := extensionsMap["timescaledb"]; ok {
+		err := enableTimescaleDB(ctx, iof)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "cannot add timescaldb to config", err)
+		}
+	}
+
+	cluster := &stackgresv1.SGCluster{}
+	err = iof.Desired.GetFromObject(ctx, cluster, "cluster")
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "not able to get cluster", err)
+	}
+
+	finalExtensions := []stackgresv1.SGClusterSpecPostgresExtensionsItem{}
+
+	for _, ext := range extensionsMap {
+		finalExtensions = append(finalExtensions, ext)
+	}
+
+	cluster.Spec.Postgres.Extensions = &finalExtensions
```

Comment:
The order when iterating over a map is not fixed. So this `extensions` array will have a different order every time.  Did you check that we do not introduce unnecessary updates of the cluster spec every reconcile loop?

Reply:
Good point, I did not check that. But I think with a sort before we apply it, it should be a non-issue.